[aarch64] Fix aarch64 build so that quantize_val_arm is defined

psaab opened the pull request:
Summary: quantize_val_arm is used in the kernels when building under aarch64

Test Plan: CI

Differential Revision: D39272746

kimishpatel commented on aten/src/ATen/native/quantized/AffineQuantizerBase.cpp:
```diff
@@ -33,7 +33,7 @@ void checkZeroPoint(const std::string& fn_name, int64_t zero_point) {
 } // anonymous namespace
 
-#ifdef USE_FBGEMM
+#if defined(USE_FBGEMM) && !defined(__aarch64__)
```
Why is this qualification needed?

psaab commented:
Otherwise in aarch64 builds quantize_val_arm doesn't get defined because it
isn't in the fbgemm case. If there's a better way to deal with this I'm all
for it.

On Tue, Sep 6, 2022 at 8:00 AM Kimish Patel ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In aten/src/ATen/native/quantized/AffineQuantizerBase.cpp
> <https://github.com/pytorch/pytorch/pull/84564#discussion_r963819589>:
>
> > @@ -33,7 +33,7 @@ void checkZeroPoint(const std::string& fn_name, int64_t zero_point) {
>
>  } // anonymous namespace
>
> -#ifdef USE_FBGEMM
> +#if defined(USE_FBGEMM) && !defined(__aarch64__)
>
> Why is this qualification needed?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/84564#pullrequestreview-1097757291>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABQCUVF2GII66J2NS3NU3TV45L7XANCNFSM6AAAAAAQFZR4GI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

kimishpatel commented:
> quantize_val_arm

I dont follow. quantize_val_arm is defined "if you are not USE_FBGEMM". Your changes say define quantize_val_arm "if you are not USE_FBGEMM OR you are aarch64". But if you are building for aarch64 you shouldnt really have USE_FBGEMM=1. SO not sure why the build would fail.

quantize_val_arm is actually used here https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L3373. So if you want to use appropriate macros to define it, a better thing to do would be to refactor that such that quantize_val_arm uses this, https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L3274.

I do agree though that this part of the code can be refactored for better readability. It is a bit of a mess.

psaab commented:
I took the easy way out to get things working. I'll refactor. Thx

On Tue, Sep 6, 2022 at 8:33 AM Kimish Patel ***@***.***>
wrote:

> quantize_val_arm
>
> I dont follow. quantize_val_arm is defined "if you are not USE_FBGEMM".
> Your changes say define quantize_val_arm "if you are not USE_FBGEMM OR you
> are aarch64". But if you are building for aarch64 you shouldnt really have
> USE_FBGEMM=1. SO not sure why the build would fail.
>
> quantize_val_arm is actually used here
> https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L3373.
> So if you want to use appropriate macros to define it, a better thing to do
> would be to refactor that such that quantize_val_arm uses this,
> https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L3274
> .
>
> I do agree though that this part of the code can be refactored for better
> readability. It is a bit of a mess.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/84564#issuecomment-1238314956>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABQCUVJQJTQFBAK2EK3SF3V45P27ANCNFSM6AAAAAAQFZR4GI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

kimishpatel commented:
> I took the easy way out to get things working. I'll refactor. Thx
> […](#)
> On Tue, Sep 6, 2022 at 8:33 AM Kimish Patel ***@***.***> wrote: quantize_val_arm I dont follow. quantize_val_arm is defined "if you are not USE_FBGEMM". Your changes say define quantize_val_arm "if you are not USE_FBGEMM OR you are aarch64". But if you are building for aarch64 you shouldnt really have USE_FBGEMM=1. SO not sure why the build would fail. quantize_val_arm is actually used here https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L3373. So if you want to use appropriate macros to define it, a better thing to do would be to refactor that such that quantize_val_arm uses this, https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L3274 . I do agree though that this part of the code can be refactored for better readability. It is a bit of a mess. — Reply to this email directly, view it on GitHub <[#84564 (comment)](https://github.com/pytorch/pytorch/pull/84564#issuecomment-1238314956)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AABQCUVJQJTQFBAK2EK3SF3V45P27ANCNFSM6AAAAAAQFZR4GI> . You are receiving this because you authored the thread.Message ID: ***@***.***>

Sure but were you building for aarch64 with `USE_FBGEMM=1`? I dont think you want to do that. And if you dont do that, it should likely just work out of the box.

psaab commented:
Well, we're trying to actually use simde to wrap a lot of the AVX/SSE code for aarch64 vs do a direct aarch64 port.

jianyuh commented:
I guess ideally we would have QNNPACK for arm on mobile side for PyTorch, but @psaab  is also exploring the option of using arm for server side (so checking if FBGEMM is working with simde conversion in FBGEMM PR like https://github.com/pytorch/FBGEMM/pull/1271 ).

kimishpatel reviewed (approved):
Looks ok for now. Do leave a comment as to a) why do this and b) maybe a better refactor might be to get _arm variants into arm specific macros.

psaab commented:
@pytorchbot merge -g

psaab commented:
@pytorchbot merge -g

psaab committed:
[aarch64] Fix aarch64 build so that quantize_val_arm is defined (#84564)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/84564

quantize_val_arm is used in the kernels when building under aarch64

Test Plan: CI

Reviewed By: kimishpatel, pallab-zz

Differential Revision: D39272746

fbshipit-source-id: 611a9a7b7f89ca268cd62a9e72db3f4f4c435fb9

ajtulloch commented:
@pytorchbot merge -g
